samza-api/src/main/java/org/apache/samza/system/StreamSpec.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.system;
+
+import java.util.Properties;
+
+
+/**
+ * StreamSpec is a blueprint for creating, validating, or simply describing a stream in the runtime environment.
+ *
+ * It has specific attributes for common behaviors that samza uses.
+ *
+ * It also includes a set of properties which may be systemName-specific.
+ */
+public class StreamSpec {
```

Comment:
The systemName would still apply for Collection streams; it would correspond to a new CollectionSystemFactory. In other words, we'd continue to leverage the System abstraction, even for trivial cases like Collections.

The physical name would be unnecessary for any local-only system. But as the existence or absence of that field would be the only difference between local and remote systems, I didn't feel that a more complicated class hierarchy was justified.

I think all the questions were answered by the above two statements. If not, let me know.